From f46436faa6d7161623ffb257c0acdd838fd7a2e7 Mon Sep 17 00:00:00 2001
From: Tobias <ruck.tobias@gmail.com>
Date: Thu, 18 Jul 2024 18:10:46 +0200
Subject: [PATCH] =?UTF-8?q?Use=20DOGE=20and=20=C2=B5DOGE=20unit=20insted?=
 =?UTF-8?q?=20of=20BCHA=20and=20XEC=20(#31)?=
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

Dogecoin has 8 decimal places by default and the DOGE ticker.

To not break all tests, we add the -usedogeunit command line flag, which is similar to -ecash, except it is on by default, and changes the units to 8 places (without affecting the address prefix), and then we disable -usedoge=0 on the test framework.

diff --git a/src/avalanche/test/processor_tests.cpp b/src/avalanche/test/processor_tests.cpp
index 1ae3c5c5e085a319b7ad6a248c941d5e3e3400e2..cbbdac9604c88ed2f4bdb0515078860393eabab6 100644
--- a/src/avalanche/test/processor_tests.cpp
+++ b/src/avalanche/test/processor_tests.cpp
@@ -1875,8 +1875,8 @@ BOOST_AUTO_TEST_CASE(quorum_detection_parameter_validation) {
 
             // Min stake is out of range
             {"-1", "0", "0", false},
-            {"-0.01", "0", "0", false},
-            {"10000000000000000.01", "0", "0", false},
+            {"-0.00000001", "0", "0", false},
+            {"10000000000.00000001", "0", "0", false},
 
             // Min connected ratio is out of range
             {"0", "-1", "0", false},
@@ -1887,12 +1887,12 @@ BOOST_AUTO_TEST_CASE(quorum_detection_parameter_validation) {
 
             // All parameters are valid
             {"0", "0", "0", true},
-            {"0.00", "0", "0", true},
-            {"0.01", "0", "0", true},
-            {"1", "0.1", "0", true},
-            {"10", "0.5", "0", true},
-            {"10", "1", "0", true},
-            {"10000000000000000.00", "0", "0", true},
+            {"0.00000000", "0", "0", true},
+            {"0.00000001", "0", "0", true},
+            {"0.00000100", "0.1", "0", true},
+            {"0.00001000", "0.5", "0", true},
+            {"0.00001000", "1", "0", true},
+            {"10000000000.00000000", "0", "0", true},
             {"0", "0", "1", true},
             {"0", "0", "100", true},
         };
diff --git a/src/cashaddr.h b/src/cashaddr.h
index 18010c33e225200c24c2c39c59fbcd069e7c6b13..68ed604541879bdc02574ea90b6eceb96c0146aa 100644
--- a/src/cashaddr.h
+++ b/src/cashaddr.h
@@ -14,6 +14,9 @@
 
 namespace cashaddr {
 
+/** Whether to enable "ecash:" addresses by default */
+constexpr bool DEFAULT_ECASH = true;
+
 /**
  * Encode a cashaddr string. Returns the empty string in case of failure.
  */
diff --git a/src/chainparams.cpp b/src/chainparams.cpp
index ae57490db0bb8cf0e0773b776b6e95ba2591e95c..2649d76ed09dc73216534af14b22c98aafb444fe 100644
--- a/src/chainparams.cpp
+++ b/src/chainparams.cpp
@@ -7,6 +7,7 @@
 #include <chainparams.h>
 
 #include <chainparamsbase.h>
+#include <cashaddr.h>
 #include <common/args.h>
 #include <consensus/merkle.h>
 #include <currencyunit.h>
@@ -25,7 +26,7 @@ const CChainParams &Params() {
 
 void ReadChainArgs(const ArgsManager &args,
                    CChainParams::ChainOptions &options) {
-    options.ecash = args.GetBoolArg("-ecash", DEFAULT_ECASH);
+    options.ecash = args.GetBoolArg("-ecash", cashaddr::DEFAULT_ECASH);
     // Only relevant for REGTEST
     options.fastprune = args.GetBoolArg("-fastprune", false);
 }
diff --git a/src/consensus/amount.cpp b/src/consensus/amount.cpp
index 85efa460a01d6941826711a9929636886e74454e..35d4b5e84005c401fd02e9837c1b0c5fbdfa6894 100644
--- a/src/consensus/amount.cpp
+++ b/src/consensus/amount.cpp
@@ -14,16 +14,17 @@
 
 using namespace std::string_view_literals;
 
-static constexpr Currency BCHA{COIN, SATOSHI, 8, "BCHA"sv};
-static constexpr Currency XEC{100 * SATOSHI, SATOSHI, 2, "XEC"sv};
+static constexpr Currency DOGE{COIN, SATOSHI, 8, "DOGE"sv};
+static constexpr Currency uDOGE{100 * SATOSHI, SATOSHI, 2, "µDOGE"sv};
 
-static_assert(BCHA.baseunit > 1 * SATOSHI,
-              "BCHA base unit must be greater than 1 satoshi");
-static_assert(XEC.baseunit > 1 * SATOSHI,
-              "XEC base unit must be greater than 1 satoshi");
+static_assert(DOGE.baseunit > 1 * SATOSHI,
+              "DOGE base unit must be greater than 1 satoshi");
+static_assert(uDOGE.baseunit > 1 * SATOSHI,
+              "uDOGE base unit must be greater than 1 satoshi");
 
 const Currency &Currency::get() {
-    return gArgs.GetBoolArg("-ecash", DEFAULT_ECASH) ? XEC : BCHA;
+    return gArgs.GetBoolArg("-usedogeunit", DEFAULT_USE_DOGE_UNIT) ? DOGE
+                                                                   : uDOGE;
 }
 
 std::string Amount::ToString() const {
diff --git a/src/currencyunit.cpp b/src/currencyunit.cpp
index 939270e89492814d185ed546726b6c58128886bc..47d657023cd0d4ee6472f518204d77a0efcdf56d 100644
--- a/src/currencyunit.cpp
+++ b/src/currencyunit.cpp
@@ -2,6 +2,7 @@
 // Distributed under the MIT software license, see the accompanying
 // file COPYING or http://www.opensource.org/licenses/mit-license.php.
 
+#include <cashaddr.h>
 #include <currencyunit.h>
 
 #include <common/args.h>
@@ -9,7 +10,11 @@
 void SetupCurrencyUnitOptions(ArgsManager &argsman) {
     // whether to use eCash default unit and address prefix
     argsman.AddArg("-ecash",
-                   strprintf("Use the eCash prefixes and units (default: %s)",
-                             DEFAULT_ECASH ? "true" : "false"),
+                   strprintf("Use the eCash address prefix (default: %s)",
+                             cashaddr::DEFAULT_ECASH ? "true" : "false"),
+                   ArgsManager::ALLOW_ANY, OptionsCategory::OPTIONS);
+    argsman.AddArg("-usedogeunit",
+                   strprintf("Use the DOGE unit (default: %s)",
+                             DEFAULT_USE_DOGE_UNIT ? "true" : "false"),
                    ArgsManager::ALLOW_ANY, OptionsCategory::OPTIONS);
 }
diff --git a/src/currencyunit.h b/src/currencyunit.h
index d0e331e2e547f3a5b97a8e5ccf4c1dd7c546d993..1b62b2fa65c854de9f94e5d8b5a9b6f93b31ac44 100644
--- a/src/currencyunit.h
+++ b/src/currencyunit.h
@@ -7,7 +7,7 @@
 
 class ArgsManager;
 
-constexpr bool DEFAULT_ECASH = true;
+constexpr bool DEFAULT_USE_DOGE_UNIT = true;
 
 void SetupCurrencyUnitOptions(ArgsManager &argsman);
 
diff --git a/src/minerfund.cpp b/src/minerfund.cpp
index d69f32a8f48ae00a4a9846b4549f3f02a21e4ee0..9b4fe774f30bf841a2e4fd231ee9c52514f8eb7f 100644
--- a/src/minerfund.cpp
+++ b/src/minerfund.cpp
@@ -5,6 +5,7 @@
 #include <minerfund.h>
 
 #include <blockindex.h>
+#include <cashaddr.h>
 #include <chainparams.h>
 #include <common/args.h>
 #include <consensus/activation.h>
@@ -40,9 +41,10 @@ static const CTxDestination &GetMinerFundDestination() {
         "ecash:prfhcnyqnl5cgrnmlfmms675w93ld7mvvqd0y8lz07";
     static const std::string bitcoinCashMinerFund =
         "bitcoincash:prfhcnyqnl5cgrnmlfmms675w93ld7mvvq5zsvycff";
-    static CTxDestination dest = BuildDestination(
-        gArgs.GetBoolArg("-ecash", DEFAULT_ECASH) ? ecashMinerFund
-                                                  : bitcoinCashMinerFund);
+    static CTxDestination dest =
+        BuildDestination(gArgs.GetBoolArg("-ecash", cashaddr::DEFAULT_ECASH)
+                             ? ecashMinerFund
+                             : bitcoinCashMinerFund);
 
     return dest;
 }
diff --git a/src/qt/bitcoinunits.cpp b/src/qt/bitcoinunits.cpp
index 0c1277996838ec31b9a65a6cc64c92fdc4236edc..91ee137aebad2d8ab69fd77912832655c7b2524f 100644
--- a/src/qt/bitcoinunits.cpp
+++ b/src/qt/bitcoinunits.cpp
@@ -21,26 +21,27 @@ using unitNameMap =
             QString /* description */
         >
     >;
-static const unitNameMap xecUnits = {
+static const unitNameMap dogeUnits = {
     {BitcoinUnits::Unit::base,
-        {"XEC",
-        "eCash"}},
+        {"DOGE",
+         "Dogecoin"}},
     {BitcoinUnits::Unit::sub,
         {"Satoshi (sat)",
-        "Satoshi (sat) (1 / 100)"}},
+         "Satoshi (sat) (1 / 100" THIN_SP_UTF8 "000" THIN_SP_UTF8 "000)"}},
 };
-static const unitNameMap bchUnits = {
+static const unitNameMap udogeUnits = {
     {BitcoinUnits::Unit::base,
-        {"BCHA",
-        "Bitcoins"}},
+        {"µDOGE",
+         "Micro Dogecoin"}},
     {BitcoinUnits::Unit::sub,
         {"Satoshi (sat)",
-        "Satoshi (sat) (1 / 100" THIN_SP_UTF8 "000" THIN_SP_UTF8 "000)"}},
+         "Satoshi (sat) (1 / 100)"}},
 };
 // clang-format on
 
 static const unitNameMap &getUnitsAtRuntime() {
-    return gArgs.GetBoolArg("-ecash", DEFAULT_ECASH) ? xecUnits : bchUnits;
+    return gArgs.GetBoolArg("-usedogeunit", DEFAULT_USE_DOGE_UNIT) ? dogeUnits
+                                                                   : udogeUnits;
 }
 
 BitcoinUnits::BitcoinUnits(QObject *parent)
diff --git a/src/qt/test/uritests.cpp b/src/qt/test/uritests.cpp
index 4bd9c64ea29536275ce24482421c5645d5961539..f2ce43d2ca394b17bdfe43249bde2363ef4e10db 100644
--- a/src/qt/test/uritests.cpp
+++ b/src/qt/test/uritests.cpp
@@ -41,7 +41,7 @@ void URITests::uriTestsCashAddr() {
     QVERIFY(rv.amount == Amount::zero());
 
     uri.setUrl(QString(
-        "ecash:qpm2qsznhks23z7629mms6s4cwef74vcwva87rkuu2?amount=0.01"));
+        "ecash:qpm2qsznhks23z7629mms6s4cwef74vcwva87rkuu2?amount=0.00000001"));
     QVERIFY(GUIUtil::parseBitcoinURI(scheme, uri, &rv));
     QVERIFY(rv.address ==
             QString("ecash:qpm2qsznhks23z7629mms6s4cwef74vcwva87rkuu2"));
@@ -49,16 +49,15 @@ void URITests::uriTestsCashAddr() {
     QVERIFY(rv.amount == SATOSHI);
 
     uri.setUrl(QString(
-        "ecash:qpm2qsznhks23z7629mms6s4cwef74vcwva87rkuu2?amount=10.01"));
+        "ecash:qpm2qsznhks23z7629mms6s4cwef74vcwva87rkuu2?amount=0.00001001"));
     QVERIFY(GUIUtil::parseBitcoinURI(scheme, uri, &rv));
     QVERIFY(rv.address ==
             QString("ecash:qpm2qsznhks23z7629mms6s4cwef74vcwva87rkuu2"));
     QVERIFY(rv.label == QString());
     QVERIFY(rv.amount == 1001 * SATOSHI);
 
-    uri.setUrl(
-        QString("ecash:qpm2qsznhks23z7629mms6s4cwef74vcwva87rkuu2?amount=100&"
-                "label=Wikipedia Example"));
+    uri.setUrl(QString("ecash:qpm2qsznhks23z7629mms6s4cwef74vcwva87rkuu2?"
+                       "amount=0.00010000&label=Wikipedia Example"));
     QVERIFY(GUIUtil::parseBitcoinURI(scheme, uri, &rv));
     QVERIFY(rv.address ==
             QString("ecash:qpm2qsznhks23z7629mms6s4cwef74vcwva87rkuu2"));
diff --git a/src/test/amount_tests.cpp b/src/test/amount_tests.cpp
index 8de0e18913bef845c8a62f11c198b130467cc7cc..5ffe880579f6c3269e24107a408b745abbb5fed7 100644
--- a/src/test/amount_tests.cpp
+++ b/src/test/amount_tests.cpp
@@ -163,7 +163,7 @@ BOOST_AUTO_TEST_CASE(ToStringTest) {
     CFeeRate feeRate;
     feeRate = CFeeRate(1 * SATOSHI);
     BOOST_CHECK_EQUAL(feeRate.ToString(),
-                      "0.01 " + Currency::getTicker() + "/kB");
+                      "0.00000001 " + Currency::getTicker() + "/kB");
 }
 
 BOOST_AUTO_TEST_SUITE_END()
diff --git a/src/test/data/script_tests.json b/src/test/data/script_tests.json
index 93bb3fa9408a5c7258630bce70eeb144de278343..28c2b83ce799677af74e8780135c884b85c4f13d 100644
--- a/src/test/data/script_tests.json
+++ b/src/test/data/script_tests.json
@@ -2915,7 +2915,7 @@
 ],
 [
     [
-        123450000000.00
+        123450.00000000
     ],
     "0x47 0x304402206e3afa6dd4d1db87538fa48a0ef3f824d7ec554103fe5fb3527254d78bd79617022066097e981df0d1ce3c07224a73585ca1f5e8ebdf79639b2cc73c20bde066075141",
     "0x41 0x0479be667ef9dcbbac55a06295ce870b07029bfcdb2dce28d959f2815b16f81798483ada7726a3c4655da4fbfc0e1108a8fd17b448a68554199c47d08ffb10d4b8 CHECKSIG",
@@ -2925,7 +2925,7 @@
 ],
 [
     [
-        123450000000.00
+        123450.00000000
     ],
     "0x47 0x3044022054e19540c81ccb214b62acda158faf01440d98adab2e13ed43a4d35cc551bfff02205b05fc09882819b3fdbe56adfef63243d7faa5c4837db0c72ad9356eae9a5b8c41",
     "0x41 0x0479be667ef9dcbbac55a06295ce870b07029bfcdb2dce28d959f2815b16f81798483ada7726a3c4655da4fbfc0e1108a8fd17b448a68554199c47d08ffb10d4b8 CHECKSIG",
@@ -2935,7 +2935,7 @@
 ],
 [
     [
-        123450000000.00
+        123450.00000000
     ],
     "0x47 0x304402206e3afa6dd4d1db87538fa48a0ef3f824d7ec554103fe5fb3527254d78bd79617022066097e981df0d1ce3c07224a73585ca1f5e8ebdf79639b2cc73c20bde066075141",
     "0x41 0x0479be667ef9dcbbac55a06295ce870b07029bfcdb2dce28d959f2815b16f81798483ada7726a3c4655da4fbfc0e1108a8fd17b448a68554199c47d08ffb10d4b8 CHECKSIG",
@@ -2945,7 +2945,7 @@
 ],
 [
     [
-        123450000000.00
+        123450.00000000
     ],
     "0x47 0x304402205ce75e7249823fe2cc78f861579189f8e10585008aa057d4e6d755bfca1c8c6802200f63547feabc3f198f18742e84d57d6ca4bdd477d10a121554647cba36742f2b41",
     "0x41 0x0479be667ef9dcbbac55a06295ce870b07029bfcdb2dce28d959f2815b16f81798483ada7726a3c4655da4fbfc0e1108a8fd17b448a68554199c47d08ffb10d4b8 CHECKSIG",
@@ -2955,7 +2955,7 @@
 ],
 [
     [
-        123450000000.00
+        123450.00000000
     ],
     "0x47 0x304402206e3afa6dd4d1db87538fa48a0ef3f824d7ec554103fe5fb3527254d78bd79617022066097e981df0d1ce3c07224a73585ca1f5e8ebdf79639b2cc73c20bde066075141",
     "0x41 0x0479be667ef9dcbbac55a06295ce870b07029bfcdb2dce28d959f2815b16f81798483ada7726a3c4655da4fbfc0e1108a8fd17b448a68554199c47d08ffb10d4b8 CHECKSIG",
diff --git a/src/test/rpc_tests.cpp b/src/test/rpc_tests.cpp
index 64f334ee68b82fdcc9c0029353de638d3c23058d..4cb432076eee545794adef1e4ca1a14f1d19f2b8 100644
--- a/src/test/rpc_tests.cpp
+++ b/src/test/rpc_tests.cpp
@@ -381,66 +381,66 @@ BOOST_AUTO_TEST_CASE(rpc_createraw_op_return) {
 }
 
 BOOST_AUTO_TEST_CASE(rpc_format_monetary_values) {
-    BOOST_CHECK(UniValue(Amount::zero()).write() == "0.00");
-    BOOST_CHECK(UniValue(SATOSHI).write() == "0.01");
-    BOOST_CHECK(UniValue(17622195 * SATOSHI).write() == "176221.95");
-    BOOST_CHECK(UniValue(50000000 * SATOSHI).write() == "500000.00");
-    BOOST_CHECK(UniValue(89898989 * SATOSHI).write() == "898989.89");
-    BOOST_CHECK(UniValue(100000000 * SATOSHI).write() == "1000000.00");
-    BOOST_CHECK(UniValue(int64_t(2099999999999990) * SATOSHI).write() ==
-                "20999999999999.90");
-    BOOST_CHECK(UniValue(int64_t(2099999999999999) * SATOSHI).write() ==
-                "20999999999999.99");
-
-    BOOST_CHECK_EQUAL(UniValue(Amount::zero()).write(), "0.00");
+    BOOST_CHECK_EQUAL(UniValue(Amount::zero()).write(), "0.00000000");
+    BOOST_CHECK_EQUAL(UniValue(SATOSHI).write(), "0.00000001");
+    BOOST_CHECK_EQUAL(UniValue(17622195 * SATOSHI).write(), "0.17622195");
+    BOOST_CHECK_EQUAL(UniValue(50000000 * SATOSHI).write(), "0.50000000");
+    BOOST_CHECK_EQUAL(UniValue(89898989 * SATOSHI).write(), "0.89898989");
+    BOOST_CHECK_EQUAL(UniValue(100000000 * SATOSHI).write(), "1.00000000");
+    BOOST_CHECK_EQUAL(UniValue(int64_t(2099999999999990) * SATOSHI).write(),
+                      "20999999.99999990");
+    BOOST_CHECK_EQUAL(UniValue(int64_t(2099999999999999) * SATOSHI).write(),
+                      "20999999.99999999");
+
+    BOOST_CHECK_EQUAL(UniValue(Amount::zero()).write(), "0.00000000");
     BOOST_CHECK_EQUAL(UniValue(123456789 * (COIN / 10000)).write(),
-                      "12345678900.00");
-    BOOST_CHECK_EQUAL(UniValue(-1 * COIN).write(), "-1000000.00");
-    BOOST_CHECK_EQUAL(UniValue(-1 * COIN / 10).write(), "-100000.00");
-
-    BOOST_CHECK_EQUAL(UniValue(100000000 * COIN).write(), "100000000000000.00");
-    BOOST_CHECK_EQUAL(UniValue(10000000 * COIN).write(), "10000000000000.00");
-    BOOST_CHECK_EQUAL(UniValue(1000000 * COIN).write(), "1000000000000.00");
-    BOOST_CHECK_EQUAL(UniValue(100000 * COIN).write(), "100000000000.00");
-    BOOST_CHECK_EQUAL(UniValue(10000 * COIN).write(), "10000000000.00");
-    BOOST_CHECK_EQUAL(UniValue(1000 * COIN).write(), "1000000000.00");
-    BOOST_CHECK_EQUAL(UniValue(100 * COIN).write(), "100000000.00");
-    BOOST_CHECK_EQUAL(UniValue(10 * COIN).write(), "10000000.00");
-    BOOST_CHECK_EQUAL(UniValue(COIN).write(), "1000000.00");
-    BOOST_CHECK_EQUAL(UniValue(COIN / 10).write(), "100000.00");
-    BOOST_CHECK_EQUAL(UniValue(COIN / 100).write(), "10000.00");
-    BOOST_CHECK_EQUAL(UniValue(COIN / 1000).write(), "1000.00");
-    BOOST_CHECK_EQUAL(UniValue(COIN / 10000).write(), "100.00");
-    BOOST_CHECK_EQUAL(UniValue(COIN / 100000).write(), "10.00");
-    BOOST_CHECK_EQUAL(UniValue(COIN / 1000000).write(), "1.00");
-    BOOST_CHECK_EQUAL(UniValue(COIN / 10000000).write(), "0.10");
-    BOOST_CHECK_EQUAL(UniValue(COIN / 100000000).write(), "0.01");
+                      "12345.67890000");
+    BOOST_CHECK_EQUAL(UniValue(-1 * COIN).write(), "-1.00000000");
+    BOOST_CHECK_EQUAL(UniValue(-1 * COIN / 10).write(), "-0.10000000");
+
+    BOOST_CHECK_EQUAL(UniValue(100000000 * COIN).write(), "100000000.00000000");
+    BOOST_CHECK_EQUAL(UniValue(10000000 * COIN).write(), "10000000.00000000");
+    BOOST_CHECK_EQUAL(UniValue(1000000 * COIN).write(), "1000000.00000000");
+    BOOST_CHECK_EQUAL(UniValue(100000 * COIN).write(), "100000.00000000");
+    BOOST_CHECK_EQUAL(UniValue(10000 * COIN).write(), "10000.00000000");
+    BOOST_CHECK_EQUAL(UniValue(1000 * COIN).write(), "1000.00000000");
+    BOOST_CHECK_EQUAL(UniValue(100 * COIN).write(), "100.00000000");
+    BOOST_CHECK_EQUAL(UniValue(10 * COIN).write(), "10.00000000");
+    BOOST_CHECK_EQUAL(UniValue(COIN).write(), "1.00000000");
+    BOOST_CHECK_EQUAL(UniValue(COIN / 10).write(), "0.10000000");
+    BOOST_CHECK_EQUAL(UniValue(COIN / 100).write(), "0.01000000");
+    BOOST_CHECK_EQUAL(UniValue(COIN / 1000).write(), "0.00100000");
+    BOOST_CHECK_EQUAL(UniValue(COIN / 10000).write(), "0.00010000");
+    BOOST_CHECK_EQUAL(UniValue(COIN / 100000).write(), "0.00001000");
+    BOOST_CHECK_EQUAL(UniValue(COIN / 1000000).write(), "0.00000100");
+    BOOST_CHECK_EQUAL(UniValue(COIN / 10000000).write(), "0.00000010");
+    BOOST_CHECK_EQUAL(UniValue(COIN / 100000000).write(), "0.00000001");
 
     BOOST_CHECK_EQUAL(
         UniValue(std::numeric_limits<int64_t>::max() * SATOSHI).write(),
-        "92233720368547758.07");
+        "92233720368.54775807");
     BOOST_CHECK_EQUAL(
         UniValue((std::numeric_limits<int64_t>::max() - 1) * SATOSHI).write(),
-        "92233720368547758.06");
+        "92233720368.54775806");
     BOOST_CHECK_EQUAL(
         UniValue((std::numeric_limits<int64_t>::max() - 2) * SATOSHI).write(),
-        "92233720368547758.05");
+        "92233720368.54775805");
     BOOST_CHECK_EQUAL(
         UniValue((std::numeric_limits<int64_t>::max() - 3) * SATOSHI).write(),
-        "92233720368547758.04");
+        "92233720368.54775804");
     // ...
     BOOST_CHECK_EQUAL(
         UniValue((std::numeric_limits<int64_t>::min() + 3) * SATOSHI).write(),
-        "-92233720368547758.05");
+        "-92233720368.54775805");
     BOOST_CHECK_EQUAL(
         UniValue((std::numeric_limits<int64_t>::min() + 2) * SATOSHI).write(),
-        "-92233720368547758.06");
+        "-92233720368.54775806");
     BOOST_CHECK_EQUAL(
         UniValue((std::numeric_limits<int64_t>::min() + 1) * SATOSHI).write(),
-        "-92233720368547758.07");
+        "-92233720368.54775807");
     BOOST_CHECK_EQUAL(
         UniValue((std::numeric_limits<int64_t>::min()) * SATOSHI).write(),
-        "-92233720368547758.08");
+        "-92233720368.54775808");
 }
 
 static UniValue ValueFromString(const std::string &str) noexcept {
@@ -453,46 +453,43 @@ BOOST_AUTO_TEST_CASE(rpc_parse_monetary_values) {
     BOOST_CHECK_THROW(AmountFromValue(ValueFromString("-0.01")), UniValue);
     BOOST_CHECK_EQUAL(AmountFromValue(ValueFromString("0")), Amount::zero());
     BOOST_CHECK_EQUAL(AmountFromValue(ValueFromString("0.00")), Amount::zero());
-    BOOST_CHECK_EQUAL(AmountFromValue(ValueFromString("0.01")), SATOSHI);
+    BOOST_CHECK_EQUAL(AmountFromValue(ValueFromString("0.01")), CENT);
     BOOST_CHECK_EQUAL(AmountFromValue(ValueFromString("176221.95")),
-                      17622195 * SATOSHI);
+                      17622195 * CENT);
     BOOST_CHECK_EQUAL(AmountFromValue(ValueFromString("500000")),
-                      50000000 * SATOSHI);
+                      50000000 * CENT);
     BOOST_CHECK_EQUAL(AmountFromValue(ValueFromString("500000.00")),
-                      50000000 * SATOSHI);
+                      50000000 * CENT);
     BOOST_CHECK_EQUAL(AmountFromValue(ValueFromString("898989.89")),
-                      89898989 * SATOSHI);
+                      89898989 * CENT);
     BOOST_CHECK_EQUAL(AmountFromValue(ValueFromString("1000000.00")),
-                      100000000 * SATOSHI);
-    BOOST_CHECK_EQUAL(AmountFromValue(ValueFromString("20999999999999.9")),
+                      100000000 * CENT);
+    BOOST_CHECK_EQUAL(AmountFromValue(ValueFromString("20999999.9999999")),
                       int64_t(2099999999999990) * SATOSHI);
-    BOOST_CHECK_EQUAL(AmountFromValue(ValueFromString("20999999999999.99")),
+    BOOST_CHECK_EQUAL(AmountFromValue(ValueFromString("20999999.99999999")),
                       int64_t(2099999999999999) * SATOSHI);
 
-    BOOST_CHECK_EQUAL(AmountFromValue(ValueFromString("1e-2")),
-                      COIN / 100000000);
-    BOOST_CHECK_EQUAL(AmountFromValue(ValueFromString("0.1e-1")),
-                      COIN / 100000000);
-    BOOST_CHECK_EQUAL(AmountFromValue(ValueFromString("0.01e-0")),
-                      COIN / 100000000);
+    BOOST_CHECK_EQUAL(AmountFromValue(ValueFromString("1e-2")), CENT);
+    BOOST_CHECK_EQUAL(AmountFromValue(ValueFromString("0.1e-1")), CENT);
+    BOOST_CHECK_EQUAL(AmountFromValue(ValueFromString("0.01e-0")), CENT);
     BOOST_CHECK_EQUAL(AmountFromValue(ValueFromString(
                           "0.00000000000000000000000000000000000001e+36")),
-                      SATOSHI);
+                      CENT);
     BOOST_CHECK_EQUAL(AmountFromValue(ValueFromString(
                           "0."
                           "0000000000000000000000000000000000000000000000000000"
                           "000000000000000000000001e+74")),
-                      COIN / 100000000);
+                      CENT);
     BOOST_CHECK_EQUAL(
         AmountFromValue(ValueFromString("10000000000000000000000000000000000000"
                                         "000000000000000000000000000e-58")),
-        COIN);
+        1000000 * COIN);
     BOOST_CHECK_EQUAL(
         AmountFromValue(ValueFromString(
             "0."
             "000000000000000000000000000000000000000000000000000000000000000100"
             "000000000000000000000000000000000000000000000000000e70")),
-        COIN);
+        1000000 * COIN);
 
     // should fail
     BOOST_CHECK_THROW(AmountFromValue(ValueFromString("1e-9")), UniValue);
@@ -500,14 +497,14 @@ BOOST_AUTO_TEST_CASE(rpc_parse_monetary_values) {
     BOOST_CHECK_THROW(AmountFromValue(ValueFromString("0.000000019")),
                       UniValue);
     // should pass, cut trailing 0
-    BOOST_CHECK_EQUAL(AmountFromValue(ValueFromString("0.01000000")), SATOSHI);
+    BOOST_CHECK_EQUAL(AmountFromValue(ValueFromString("0.00000001000000")),
+                      SATOSHI);
     // should fail
     BOOST_CHECK_THROW(AmountFromValue(ValueFromString("19e-9")), UniValue);
     // should pass, leading 0 is present
-    BOOST_CHECK_EQUAL(AmountFromValue(ValueFromString("0.19e-0")),
-                      19 * SATOSHI);
+    BOOST_CHECK_EQUAL(AmountFromValue(ValueFromString("0.19e-0")), 19 * CENT);
     // should fail, no leading 0
-    BOOST_CHECK_EXCEPTION(AmountFromValue(".19e-6"), UniValue,
+    BOOST_CHECK_EXCEPTION(AmountFromValue(".19e-9"), UniValue,
                           HasJSON(R"({"code":-3,"message":"Invalid amount"})"));
     // overflow error
     BOOST_CHECK_THROW(AmountFromValue(ValueFromString("92233720368.54775808")),
diff --git a/src/test/util_tests.cpp b/src/test/util_tests.cpp
index 30d8d96524df976912251ba1eea3f83a4d3cfe8b..d10df2f9903a5c8f7e6ae76bd7597f2c8f8a4272 100644
--- a/src/test/util_tests.cpp
+++ b/src/test/util_tests.cpp
@@ -1648,53 +1648,52 @@ BOOST_AUTO_TEST_CASE(util_ReadWriteSettings) {
 
 BOOST_AUTO_TEST_CASE(util_FormatMoney) {
     BOOST_CHECK_EQUAL(FormatMoney(Amount::zero()), "0.00");
-    BOOST_CHECK_EQUAL(FormatMoney(123456789 * (COIN / 10000)),
-                      "12345678900.00");
-    BOOST_CHECK_EQUAL(FormatMoney(-1 * COIN), "-1000000.00");
-
-    BOOST_CHECK_EQUAL(FormatMoney(100000000 * COIN), "100000000000000.00");
-    BOOST_CHECK_EQUAL(FormatMoney(10000000 * COIN), "10000000000000.00");
-    BOOST_CHECK_EQUAL(FormatMoney(1000000 * COIN), "1000000000000.00");
-    BOOST_CHECK_EQUAL(FormatMoney(100000 * COIN), "100000000000.00");
-    BOOST_CHECK_EQUAL(FormatMoney(10000 * COIN), "10000000000.00");
-    BOOST_CHECK_EQUAL(FormatMoney(1000 * COIN), "1000000000.00");
-    BOOST_CHECK_EQUAL(FormatMoney(100 * COIN), "100000000.00");
-    BOOST_CHECK_EQUAL(FormatMoney(10 * COIN), "10000000.00");
-    BOOST_CHECK_EQUAL(FormatMoney(COIN), "1000000.00");
-    BOOST_CHECK_EQUAL(FormatMoney(COIN / 10), "100000.00");
-    BOOST_CHECK_EQUAL(FormatMoney(COIN / 100), "10000.00");
-    BOOST_CHECK_EQUAL(FormatMoney(COIN / 1000), "1000.00");
-    BOOST_CHECK_EQUAL(FormatMoney(COIN / 10000), "100.00");
-    BOOST_CHECK_EQUAL(FormatMoney(COIN / 100000), "10.00");
-    BOOST_CHECK_EQUAL(FormatMoney(COIN / 1000000), "1.00");
-    BOOST_CHECK_EQUAL(FormatMoney(COIN / 10000000), "0.10");
-    BOOST_CHECK_EQUAL(FormatMoney(COIN / 100000000), "0.01");
+    BOOST_CHECK_EQUAL(FormatMoney(123456789 * (COIN / 10000)), "12345.6789");
+    BOOST_CHECK_EQUAL(FormatMoney(-1 * COIN), "-1.00");
+
+    BOOST_CHECK_EQUAL(FormatMoney(100000000 * COIN), "100000000.00");
+    BOOST_CHECK_EQUAL(FormatMoney(10000000 * COIN), "10000000.00");
+    BOOST_CHECK_EQUAL(FormatMoney(1000000 * COIN), "1000000.00");
+    BOOST_CHECK_EQUAL(FormatMoney(100000 * COIN), "100000.00");
+    BOOST_CHECK_EQUAL(FormatMoney(10000 * COIN), "10000.00");
+    BOOST_CHECK_EQUAL(FormatMoney(1000 * COIN), "1000.00");
+    BOOST_CHECK_EQUAL(FormatMoney(100 * COIN), "100.00");
+    BOOST_CHECK_EQUAL(FormatMoney(10 * COIN), "10.00");
+    BOOST_CHECK_EQUAL(FormatMoney(COIN), "1.00");
+    BOOST_CHECK_EQUAL(FormatMoney(COIN / 10), "0.10");
+    BOOST_CHECK_EQUAL(FormatMoney(COIN / 100), "0.01");
+    BOOST_CHECK_EQUAL(FormatMoney(COIN / 1000), "0.001");
+    BOOST_CHECK_EQUAL(FormatMoney(COIN / 10000), "0.0001");
+    BOOST_CHECK_EQUAL(FormatMoney(COIN / 100000), "0.00001");
+    BOOST_CHECK_EQUAL(FormatMoney(COIN / 1000000), "0.000001");
+    BOOST_CHECK_EQUAL(FormatMoney(COIN / 10000000), "0.0000001");
+    BOOST_CHECK_EQUAL(FormatMoney(COIN / 100000000), "0.00000001");
 
     BOOST_CHECK_EQUAL(
         FormatMoney(std::numeric_limits<int64_t>::max() * SATOSHI),
-        "92233720368547758.07");
+        "92233720368.54775807");
     BOOST_CHECK_EQUAL(
         FormatMoney((std::numeric_limits<int64_t>::max() - 1) * SATOSHI),
-        "92233720368547758.06");
+        "92233720368.54775806");
     BOOST_CHECK_EQUAL(
         FormatMoney((std::numeric_limits<int64_t>::max() - 2) * SATOSHI),
-        "92233720368547758.05");
+        "92233720368.54775805");
     BOOST_CHECK_EQUAL(
         FormatMoney((std::numeric_limits<int64_t>::max() - 3) * SATOSHI),
-        "92233720368547758.04");
+        "92233720368.54775804");
     // ...
     BOOST_CHECK_EQUAL(
         FormatMoney((std::numeric_limits<int64_t>::min() + 3) * SATOSHI),
-        "-92233720368547758.05");
+        "-92233720368.54775805");
     BOOST_CHECK_EQUAL(
         FormatMoney((std::numeric_limits<int64_t>::min() + 2) * SATOSHI),
-        "-92233720368547758.06");
+        "-92233720368.54775806");
     BOOST_CHECK_EQUAL(
         FormatMoney((std::numeric_limits<int64_t>::min() + 1) * SATOSHI),
-        "-92233720368547758.07");
+        "-92233720368.54775807");
     BOOST_CHECK_EQUAL(
         FormatMoney(std::numeric_limits<int64_t>::min() * SATOSHI),
-        "-92233720368547758.08");
+        "-92233720368.54775808");
 }
 
 BOOST_AUTO_TEST_CASE(util_ParseMoney) {
@@ -1703,9 +1702,9 @@ BOOST_AUTO_TEST_CASE(util_ParseMoney) {
     BOOST_CHECK_EQUAL(ret, Amount::zero());
 
     BOOST_CHECK(ParseMoney("1234567.89", ret));
-    BOOST_CHECK_EQUAL(ret, 123456789 * SATOSHI);
+    BOOST_CHECK_EQUAL(ret, int64_t(123456789000000LL) * SATOSHI);
 
-    BOOST_CHECK(ParseMoney("10000000000000000.00", ret));
+    BOOST_CHECK(ParseMoney("10000000000.00000000", ret));
     BOOST_CHECK_EQUAL(ret, MAX_MONEY);
 
     const auto XEC = Currency::get().baseunit;
@@ -1747,13 +1746,13 @@ BOOST_AUTO_TEST_CASE(util_ParseMoney) {
     BOOST_CHECK_EQUAL(ret, XEC / 100);
 
     // Parsing amount that can not be represented in ret should fail
-    BOOST_CHECK(!ParseMoney("0.001", ret));
-    BOOST_CHECK(!ParseMoney("0.0001", ret));
-    BOOST_CHECK(!ParseMoney("0.00001", ret));
-    BOOST_CHECK(!ParseMoney("0.000001", ret));
-    BOOST_CHECK(!ParseMoney("0.0000001", ret));
-    BOOST_CHECK(!ParseMoney("0.00000001", ret));
     BOOST_CHECK(!ParseMoney("0.000000001", ret));
+    BOOST_CHECK(!ParseMoney("0.0000000001", ret));
+    BOOST_CHECK(!ParseMoney("0.00000000001", ret));
+    BOOST_CHECK(!ParseMoney("0.000000000001", ret));
+    BOOST_CHECK(!ParseMoney("0.0000000000001", ret));
+    BOOST_CHECK(!ParseMoney("0.00000000000001", ret));
+    BOOST_CHECK(!ParseMoney("0.000000000000001", ret));
 
     // Parsing empty string should fail
     BOOST_CHECK(!ParseMoney("", ret));
@@ -1767,13 +1766,13 @@ BOOST_AUTO_TEST_CASE(util_ParseMoney) {
     BOOST_CHECK(!ParseMoney(" 1 2.3 ", ret));
 
     // Max possible value with the max number of digits
-    BOOST_CHECK(ParseMoney("9999999999999999.99", ret));
+    BOOST_CHECK(ParseMoney("9999999999.99999999", ret));
     BOOST_CHECK_EQUAL(ret, int64_t(999999999999999999) * SATOSHI);
-    BOOST_CHECK(ParseMoney("19999999999999999.99", ret));
+    BOOST_CHECK(ParseMoney("19999999999.99999999", ret));
     BOOST_CHECK_EQUAL(ret, int64_t(1999999999999999999) * SATOSHI);
     // Attempted 63 bit overflow should fail
-    BOOST_CHECK(!ParseMoney("20000000000000000.00", ret));
-    BOOST_CHECK(!ParseMoney("92233720368547758.08", ret));
+    BOOST_CHECK(!ParseMoney("20000000000.00000000", ret));
+    BOOST_CHECK(!ParseMoney("92233720368.54775808", ret));
 
     // Parsing negative amounts must fail
     BOOST_CHECK(!ParseMoney("-1", ret));
diff --git a/test/functional/abc_rpc_ecash.py b/test/functional/abc_rpc_ecash.py
index 07b183a1ef1a1530516ae3952644d680985e5f92..46c8fda5a478515f2cf9b4e09939e9555b66a5ef 100644
--- a/test/functional/abc_rpc_ecash.py
+++ b/test/functional/abc_rpc_ecash.py
@@ -39,15 +39,17 @@ class ECashRPCTest(BitcoinTestFramework):
         )
 
     def run_test(self):
-        self.log.info("Test with -ecash enabled (default setting)")
-        self.test_currency(ticker="XEC", satoshis_per_unit=100, decimals=2)
+        self.log.info(
+            "Test with -usedogeunit disabled (default setting in our test framework)"
+        )
+        self.test_currency(ticker="µDOGE", satoshis_per_unit=100, decimals=2)
 
-        self.log.info("Test with -ecash disabled")
+        self.log.info("Test with -usedogeunit enabled")
         # Disable fallbackfee, because its default setting for tests
         # is adapted to XEC only.
-        # In BCHA mode, it triggers a "-fallbackfee is set very high!" error.
-        self.restart_node(0, ["-ecash=0", "-fallbackfee=0"])
-        self.test_currency(ticker="BCHA", satoshis_per_unit=100_000_000, decimals=8)
+        # In DOGE mode, it triggers a "-fallbackfee is set very high!" error.
+        self.restart_node(0, ["-usedogeunit=1", "-fallbackfee=0"])
+        self.test_currency(ticker="DOGE", satoshis_per_unit=100_000_000, decimals=8)
 
 
 if __name__ == "__main__":
diff --git a/test/functional/wallet_send.py b/test/functional/wallet_send.py
index cf17ad92ff0a557f6cebb2be5ca4b2527c5d259e..7fb55181e242925e147d521c3320cf290a4885b7 100644
--- a/test/functional/wallet_send.py
+++ b/test/functional/wallet_send.py
@@ -333,8 +333,8 @@ class WalletSendTest(BitcoinTestFramework):
             expect_error=(
                 -4,
                 (
-                    "Fee rate (1.00 XEC/kB) is lower than the minimum fee rate setting"
-                    " (10.00 XEC/kB)"
+                    "Fee rate (1.00 µDOGE/kB) is lower than the minimum fee rate setting"
+                    " (10.00 µDOGE/kB)"
                 ),
             ),
         )
diff --git a/test/util/data/tt-delin1-out.json b/test/util/data/tt-delin1-out.json
index 4983ab1211e76c7a14d030a3594e6d8c57260869..5e271e73ca55bcd4b4519176c4d88c3b25770f0f 100644
--- a/test/util/data/tt-delin1-out.json
+++ b/test/util/data/tt-delin1-out.json
@@ -188,7 +188,7 @@
     ],
     "vout": [
         {
-            "value": 1378200.00,
+            "value": 1.37820000,
             "n": 0,
             "scriptPubKey": {
                 "asm": "OP_DUP OP_HASH160 8fd139bb39ced713f231c58a4d07bf6954d1c201 OP_EQUALVERIFY OP_CHECKSIG",
@@ -201,7 +201,7 @@
             }
         },
         {
-            "value": 10000.01,
+            "value": 0.01000001,
             "n": 1,
             "scriptPubKey": {
                 "asm": "OP_DUP OP_HASH160 6c772e9cf96371bba3da8cb733da70a2fcf20078 OP_EQUALVERIFY OP_CHECKSIG",
diff --git a/test/util/data/tt-delout1-out.json b/test/util/data/tt-delout1-out.json
index 0e0ed2f8f8e2c2452f1fbd84101024545ee2161e..38300d2f243e621611655e5429c29d82ca59cbc3 100644
--- a/test/util/data/tt-delout1-out.json
+++ b/test/util/data/tt-delout1-out.json
@@ -197,7 +197,7 @@
     ],
     "vout": [
         {
-            "value": 1378200.00,
+            "value": 1.37820000,
             "n": 0,
             "scriptPubKey": {
                 "asm": "OP_DUP OP_HASH160 8fd139bb39ced713f231c58a4d07bf6954d1c201 OP_EQUALVERIFY OP_CHECKSIG",
diff --git a/test/util/data/tt-locktime317000-out.json b/test/util/data/tt-locktime317000-out.json
index bb62afe5e0832128c0890501a8850e213a35c5d8..6faf3ea2ac9b16c55c383c7ac8d0972b35ac9b8b 100644
--- a/test/util/data/tt-locktime317000-out.json
+++ b/test/util/data/tt-locktime317000-out.json
@@ -197,7 +197,7 @@
     ],
     "vout": [
         {
-            "value": 1378200.00,
+            "value": 1.37820000,
             "n": 0,
             "scriptPubKey": {
                 "asm": "OP_DUP OP_HASH160 8fd139bb39ced713f231c58a4d07bf6954d1c201 OP_EQUALVERIFY OP_CHECKSIG",
@@ -210,7 +210,7 @@
             }
         },
         {
-            "value": 10000.01,
+            "value": 0.01000001,
             "n": 1,
             "scriptPubKey": {
                 "asm": "OP_DUP OP_HASH160 6c772e9cf96371bba3da8cb733da70a2fcf20078 OP_EQUALVERIFY OP_CHECKSIG",
diff --git a/test/util/data/txcreate1.hex b/test/util/data/txcreate1.hex
index 9ec6ee3531b7b2087c983637fbd13c2705f5923a..be80f3ac19568800302fd1cab84e89e95af28da7 100644
--- a/test/util/data/txcreate1.hex
+++ b/test/util/data/txcreate1.hex
@@ -1 +1 @@
-02000000031f5c38dfcf6f1a5f5a87c416076d392c87e6d41970d5ad5e477a02d66bde97580000000000ffffffff7cca453133921c50d5025878f7f738d1df891fd359763331935784cf6b9c82bf1200000000fffffffffccd319e04a996c96cfc0bf4c07539aa90bd0b1a700ef72fae535d6504f9a6220100000000ffffffff0280a81201000000001976a9141fc11f39be1729bf973a7ab6a615ca4729d6457488ac0084d717000000001976a914f2d4db28cad6502226ee484ae24505c2885cb12d88ac00000000
+02000000031f5c38dfcf6f1a5f5a87c416076d392c87e6d41970d5ad5e477a02d66bde97580000000000ffffffff7cca453133921c50d5025878f7f738d1df891fd359763331935784cf6b9c82bf1200000000fffffffffccd319e04a996c96cfc0bf4c07539aa90bd0b1a700ef72fae535d6504f9a6220100000000ffffffff0200209bf35e1000001976a9141fc11f39be1729bf973a7ab6a615ca4729d6457488ac0000e941cc6b01001976a914f2d4db28cad6502226ee484ae24505c2885cb12d88ac00000000
diff --git a/test/util/data/txcreate1.json b/test/util/data/txcreate1.json
index 5c0c28e4bebf0e9c8363523f1209cc6b8d700bed..d4477ea0bcb74c9c618e8bbb468836f8dda14930 100644
--- a/test/util/data/txcreate1.json
+++ b/test/util/data/txcreate1.json
@@ -1,6 +1,6 @@
 {
-    "txid": "fe7d174f42dce0cffa7a527e9bc8368956057619ec817648f6138b98f2533e8f",
-    "hash": "fe7d174f42dce0cffa7a527e9bc8368956057619ec817648f6138b98f2533e8f",
+    "txid": "4bee3d047f979e638677ead060599c0bedfc9bf09c8f31c4198f3c25b15b7013",
+    "hash": "4bee3d047f979e638677ead060599c0bedfc9bf09c8f31c4198f3c25b15b7013",
     "version": 2,
     "size": 201,
     "locktime": 0,
@@ -35,7 +35,7 @@
     ],
     "vout": [
         {
-            "value": 180000.00,
+            "value": 180000.00000000,
             "n": 0,
             "scriptPubKey": {
                 "asm": "OP_DUP OP_HASH160 1fc11f39be1729bf973a7ab6a615ca4729d64574 OP_EQUALVERIFY OP_CHECKSIG",
@@ -48,7 +48,7 @@
             }
         },
         {
-            "value": 4000000.00,
+            "value": 4000000.00000000,
             "n": 1,
             "scriptPubKey": {
                 "asm": "OP_DUP OP_HASH160 f2d4db28cad6502226ee484ae24505c2885cb12d OP_EQUALVERIFY OP_CHECKSIG",
@@ -61,5 +61,5 @@
             }
         }
     ],
-    "hex": "02000000031f5c38dfcf6f1a5f5a87c416076d392c87e6d41970d5ad5e477a02d66bde97580000000000ffffffff7cca453133921c50d5025878f7f738d1df891fd359763331935784cf6b9c82bf1200000000fffffffffccd319e04a996c96cfc0bf4c07539aa90bd0b1a700ef72fae535d6504f9a6220100000000ffffffff0280a81201000000001976a9141fc11f39be1729bf973a7ab6a615ca4729d6457488ac0084d717000000001976a914f2d4db28cad6502226ee484ae24505c2885cb12d88ac00000000"
+    "hex": "02000000031f5c38dfcf6f1a5f5a87c416076d392c87e6d41970d5ad5e477a02d66bde97580000000000ffffffff7cca453133921c50d5025878f7f738d1df891fd359763331935784cf6b9c82bf1200000000fffffffffccd319e04a996c96cfc0bf4c07539aa90bd0b1a700ef72fae535d6504f9a6220100000000ffffffff0200209bf35e1000001976a9141fc11f39be1729bf973a7ab6a615ca4729d6457488ac0000e941cc6b01001976a914f2d4db28cad6502226ee484ae24505c2885cb12d88ac00000000"
 }
diff --git a/test/util/data/txcreate2.json b/test/util/data/txcreate2.json
index 6370677fdfba6704153df4ab3021b2a2bdd0e21b..a628eda227b9ca7df53b1f22fb907c2847dd1c6a 100644
--- a/test/util/data/txcreate2.json
+++ b/test/util/data/txcreate2.json
@@ -8,7 +8,7 @@
     ],
     "vout": [
         {
-            "value": 0.00,
+            "value": 0.00000000,
             "n": 0,
             "scriptPubKey": {
                 "asm": "",
diff --git a/test/util/data/txcreatedata1.hex b/test/util/data/txcreatedata1.hex
index cefd1a05a62cd816b6c7904e7e4144e3e6c5531c..36fec516f368b5cadf4a710c0a18626647ed471d 100644
--- a/test/util/data/txcreatedata1.hex
+++ b/test/util/data/txcreatedata1.hex
@@ -1 +1 @@
-02000000011f5c38dfcf6f1a5f5a87c416076d392c87e6d41970d5ad5e477a02d66bde97580000000000ffffffff0280a81201000000001976a9141fc11f39be1729bf973a7ab6a615ca4729d6457488ac0084d71700000000526a4c4f54686973204f505f52455455524e207472616e73616374696f6e206f7574707574207761732063726561746564206279206d6f646966696564206372656174657261777472616e73616374696f6e2e00000000
+02000000011f5c38dfcf6f1a5f5a87c416076d392c87e6d41970d5ad5e477a02d66bde97580000000000ffffffff0200209bf35e1000001976a9141fc11f39be1729bf973a7ab6a615ca4729d6457488ac0000e941cc6b0100526a4c4f54686973204f505f52455455524e207472616e73616374696f6e206f7574707574207761732063726561746564206279206d6f646966696564206372656174657261777472616e73616374696f6e2e00000000
diff --git a/test/util/data/txcreatedata1.json b/test/util/data/txcreatedata1.json
index 3273a9411ff1e513eb3035c84bcbc4b7a94a19d6..caedb30554adff93e65a30801e15428f03b28056 100644
--- a/test/util/data/txcreatedata1.json
+++ b/test/util/data/txcreatedata1.json
@@ -1,6 +1,6 @@
 {
-    "txid": "07894b4d12fe7853dd911402db1620920d261b9627c447f931417d330c25f06e",
-    "hash": "07894b4d12fe7853dd911402db1620920d261b9627c447f931417d330c25f06e",
+    "txid": "385e14a819b4b72889dc6d9b7d4c573c722c9b2fe9b6305de5faa2a0344b28c1",
+    "hash": "385e14a819b4b72889dc6d9b7d4c573c722c9b2fe9b6305de5faa2a0344b28c1",
     "version": 1,
     "size": 176,
     "locktime": 0,
@@ -17,7 +17,7 @@
     ],
     "vout": [
         {
-            "value": 180000.00,
+            "value": 180000.00000000,
             "n": 0,
             "scriptPubKey": {
                 "asm": "OP_DUP OP_HASH160 1fc11f39be1729bf973a7ab6a615ca4729d64574 OP_EQUALVERIFY OP_CHECKSIG",
@@ -30,7 +30,7 @@
             }
         },
         {
-            "value": 4000000.00,
+            "value": 4000000.00000000,
             "n": 1,
             "scriptPubKey": {
                 "asm": "OP_RETURN 54686973204f505f52455455524e207472616e73616374696f6e206f7574707574207761732063726561746564206279206d6f646966696564206372656174657261777472616e73616374696f6e2e",
@@ -39,5 +39,5 @@
             }
         }
     ],
-    "hex": "01000000011f5c38dfcf6f1a5f5a87c416076d392c87e6d41970d5ad5e477a02d66bde97580000000000ffffffff0280a81201000000001976a9141fc11f39be1729bf973a7ab6a615ca4729d6457488ac0084d71700000000526a4c4f54686973204f505f52455455524e207472616e73616374696f6e206f7574707574207761732063726561746564206279206d6f646966696564206372656174657261777472616e73616374696f6e2e00000000"
+    "hex": "01000000011f5c38dfcf6f1a5f5a87c416076d392c87e6d41970d5ad5e477a02d66bde97580000000000ffffffff0200209bf35e1000001976a9141fc11f39be1729bf973a7ab6a615ca4729d6457488ac0000e941cc6b0100526a4c4f54686973204f505f52455455524e207472616e73616374696f6e206f7574707574207761732063726561746564206279206d6f646966696564206372656174657261777472616e73616374696f6e2e00000000"
 }
diff --git a/test/util/data/txcreatedata2.hex b/test/util/data/txcreatedata2.hex
index d69cf58ba14f1c276e1bc2cb662e80eca8f85482..23714ea74dad35e435e8073ba86338bc1729bb45 100644
--- a/test/util/data/txcreatedata2.hex
+++ b/test/util/data/txcreatedata2.hex
@@ -1 +1 @@
-02000000011f5c38dfcf6f1a5f5a87c416076d392c87e6d41970d5ad5e477a02d66bde97580000000000ffffffff0280a81201000000001976a9141fc11f39be1729bf973a7ab6a615ca4729d6457488ac0000000000000000526a4c4f54686973204f505f52455455524e207472616e73616374696f6e206f7574707574207761732063726561746564206279206d6f646966696564206372656174657261777472616e73616374696f6e2e00000000
+02000000011f5c38dfcf6f1a5f5a87c416076d392c87e6d41970d5ad5e477a02d66bde97580000000000ffffffff0200209bf35e1000001976a9141fc11f39be1729bf973a7ab6a615ca4729d6457488ac0000000000000000526a4c4f54686973204f505f52455455524e207472616e73616374696f6e206f7574707574207761732063726561746564206279206d6f646966696564206372656174657261777472616e73616374696f6e2e00000000
diff --git a/test/util/data/txcreatedata2.json b/test/util/data/txcreatedata2.json
index 98c1a55c4d782d8500119b70fd0307296b8b89b8..9b81bb22cb5816b5a757e660a7054bff63660b1b 100644
--- a/test/util/data/txcreatedata2.json
+++ b/test/util/data/txcreatedata2.json
@@ -1,6 +1,6 @@
 {
-    "txid": "c14b007fa3a6c1e7765919c1d14c1cfc2b8642c3a5d3be4b1fa8c4ccfec98bb0",
-    "hash": "c14b007fa3a6c1e7765919c1d14c1cfc2b8642c3a5d3be4b1fa8c4ccfec98bb0",
+    "txid": "dc4f05ad26cfda09435b95cdef11801bf04bbf0213f4bd0df5707317f08af9ef",
+    "hash": "dc4f05ad26cfda09435b95cdef11801bf04bbf0213f4bd0df5707317f08af9ef",
     "version": 2,
     "size": 176,
     "locktime": 0,
@@ -17,7 +17,7 @@
     ],
     "vout": [
         {
-            "value": 180000.00,
+            "value": 180000.00000000,
             "n": 0,
             "scriptPubKey": {
                 "asm": "OP_DUP OP_HASH160 1fc11f39be1729bf973a7ab6a615ca4729d64574 OP_EQUALVERIFY OP_CHECKSIG",
@@ -30,7 +30,7 @@
             }
         },
         {
-            "value": 0.00,
+            "value": 0.00000000,
             "n": 1,
             "scriptPubKey": {
                 "asm": "OP_RETURN 54686973204f505f52455455524e207472616e73616374696f6e206f7574707574207761732063726561746564206279206d6f646966696564206372656174657261777472616e73616374696f6e2e",
@@ -39,5 +39,5 @@
             }
         }
     ],
-    "hex": "02000000011f5c38dfcf6f1a5f5a87c416076d392c87e6d41970d5ad5e477a02d66bde97580000000000ffffffff0280a81201000000001976a9141fc11f39be1729bf973a7ab6a615ca4729d6457488ac0000000000000000526a4c4f54686973204f505f52455455524e207472616e73616374696f6e206f7574707574207761732063726561746564206279206d6f646966696564206372656174657261777472616e73616374696f6e2e00000000"
+    "hex": "02000000011f5c38dfcf6f1a5f5a87c416076d392c87e6d41970d5ad5e477a02d66bde97580000000000ffffffff0200209bf35e1000001976a9141fc11f39be1729bf973a7ab6a615ca4729d6457488ac0000000000000000526a4c4f54686973204f505f52455455524e207472616e73616374696f6e206f7574707574207761732063726561746564206279206d6f646966696564206372656174657261777472616e73616374696f6e2e00000000"
 }
diff --git a/test/util/data/txcreatedata_seq0.hex b/test/util/data/txcreatedata_seq0.hex
index 54b89d238116d0c3fab629ea1cc1b3a662dee928..69255dc84268b32d1bddc57b57914f5b680a38fb 100644
--- a/test/util/data/txcreatedata_seq0.hex
+++ b/test/util/data/txcreatedata_seq0.hex
@@ -1 +1 @@
-02000000011f5c38dfcf6f1a5f5a87c416076d392c87e6d41970d5ad5e477a02d66bde97580000000000fdffffff0180a81201000000001976a9141fc11f39be1729bf973a7ab6a615ca4729d6457488ac00000000
+02000000011f5c38dfcf6f1a5f5a87c416076d392c87e6d41970d5ad5e477a02d66bde97580000000000fdffffff0100209bf35e1000001976a9141fc11f39be1729bf973a7ab6a615ca4729d6457488ac00000000
diff --git a/test/util/data/txcreatedata_seq0.json b/test/util/data/txcreatedata_seq0.json
index b8c565b0d0dadf3f685b9573231ec3d0146b7f8c..14f03059b2174ec8fc93ad196744d2f626a9ed51 100644
--- a/test/util/data/txcreatedata_seq0.json
+++ b/test/util/data/txcreatedata_seq0.json
@@ -1,6 +1,6 @@
 {
-    "txid": "8df6ed527472542dd5e137c242a7c5a9f337ac34f7b257ae4af886aeaebb51b0",
-    "hash": "8df6ed527472542dd5e137c242a7c5a9f337ac34f7b257ae4af886aeaebb51b0",
+    "txid": "c1e8b9210c838acbd3ef821d426f52306c7843a1f0c4390fe54d79e00544d299",
+    "hash": "c1e8b9210c838acbd3ef821d426f52306c7843a1f0c4390fe54d79e00544d299",
     "version": 2,
     "size": 85,
     "locktime": 0,
@@ -17,7 +17,7 @@
     ],
     "vout": [
         {
-            "value": 180000.00,
+            "value": 180000.00000000,
             "n": 0,
             "scriptPubKey": {
                 "asm": "OP_DUP OP_HASH160 1fc11f39be1729bf973a7ab6a615ca4729d64574 OP_EQUALVERIFY OP_CHECKSIG",
@@ -30,5 +30,5 @@
             }
         }
     ],
-    "hex": "02000000011f5c38dfcf6f1a5f5a87c416076d392c87e6d41970d5ad5e477a02d66bde97580000000000fdffffff0180a81201000000001976a9141fc11f39be1729bf973a7ab6a615ca4729d6457488ac00000000"
+    "hex": "02000000011f5c38dfcf6f1a5f5a87c416076d392c87e6d41970d5ad5e477a02d66bde97580000000000fdffffff0100209bf35e1000001976a9141fc11f39be1729bf973a7ab6a615ca4729d6457488ac00000000"
 }
diff --git a/test/util/data/txcreatedata_seq1.json b/test/util/data/txcreatedata_seq1.json
index 2c78abbf4ba8e772a7b8fb452dced052dfc0e164..861c2923a43ea6996d8452ab4fa4611052ad6a00 100644
--- a/test/util/data/txcreatedata_seq1.json
+++ b/test/util/data/txcreatedata_seq1.json
@@ -26,7 +26,7 @@
     ],
     "vout": [
         {
-            "value": 180000.00,
+            "value": 0.18000000,
             "n": 0,
             "scriptPubKey": {
                 "asm": "OP_DUP OP_HASH160 1fc11f39be1729bf973a7ab6a615ca4729d64574 OP_EQUALVERIFY OP_CHECKSIG",
diff --git a/test/util/data/txcreatemultisig1.hex b/test/util/data/txcreatemultisig1.hex
index 9c00004d383d2b459f7958e2f11909cece7ee348..f36ee69a456b0368130ece7423047b52dad62b6c 100644
--- a/test/util/data/txcreatemultisig1.hex
+++ b/test/util/data/txcreatemultisig1.hex
@@ -1 +1 @@
-01000000000100e1f5050000000069522102a5613bd857b7048924264d1e70e08fb2a7e6527d32b7ab1bb993ac59964ff39721021ac43c7ff740014c3b33737ede99c967e4764553d1b2b83db77c83b8715fa72d2102df2089105c77f266fa11a9d33f05c735234075f2e8780824c6b709415f9fb48553ae00000000
+01000000000100407a10f35a000069522102a5613bd857b7048924264d1e70e08fb2a7e6527d32b7ab1bb993ac59964ff39721021ac43c7ff740014c3b33737ede99c967e4764553d1b2b83db77c83b8715fa72d2102df2089105c77f266fa11a9d33f05c735234075f2e8780824c6b709415f9fb48553ae00000000
diff --git a/test/util/data/txcreatemultisig1.json b/test/util/data/txcreatemultisig1.json
index 6516594858454ac720029c7f3e494364dfde9e62..cba563e52a2c5a4635febf6ecc6a9da809faf872 100644
--- a/test/util/data/txcreatemultisig1.json
+++ b/test/util/data/txcreatemultisig1.json
@@ -1,6 +1,6 @@
 {
-    "txid": "0d1d4edfc217d9db3ab6a9298f26a52eae3c52f55a6cb8ccbc14f7c727572894",
-    "hash": "0d1d4edfc217d9db3ab6a9298f26a52eae3c52f55a6cb8ccbc14f7c727572894",
+    "txid": "be70b847f5c97c5bb3ae72ecb80aeb80c90cabdba509d63a497416888a1f00a9",
+    "hash": "be70b847f5c97c5bb3ae72ecb80aeb80c90cabdba509d63a497416888a1f00a9",
     "version": 1,
     "size": 124,
     "locktime": 0,
@@ -8,7 +8,7 @@
     ],
     "vout": [
         {
-            "value": 1000000.00,
+            "value": 1000000.00000000,
             "n": 0,
             "scriptPubKey": {
                 "asm": "2 02a5613bd857b7048924264d1e70e08fb2a7e6527d32b7ab1bb993ac59964ff397 021ac43c7ff740014c3b33737ede99c967e4764553d1b2b83db77c83b8715fa72d 02df2089105c77f266fa11a9d33f05c735234075f2e8780824c6b709415f9fb485 3 OP_CHECKMULTISIG",
@@ -23,5 +23,5 @@
             }
         }
     ],
-    "hex": "01000000000100e1f5050000000069522102a5613bd857b7048924264d1e70e08fb2a7e6527d32b7ab1bb993ac59964ff39721021ac43c7ff740014c3b33737ede99c967e4764553d1b2b83db77c83b8715fa72d2102df2089105c77f266fa11a9d33f05c735234075f2e8780824c6b709415f9fb48553ae00000000"
+    "hex": "01000000000100407a10f35a000069522102a5613bd857b7048924264d1e70e08fb2a7e6527d32b7ab1bb993ac59964ff39721021ac43c7ff740014c3b33737ede99c967e4764553d1b2b83db77c83b8715fa72d2102df2089105c77f266fa11a9d33f05c735234075f2e8780824c6b709415f9fb48553ae00000000"
 }
diff --git a/test/util/data/txcreatemultisig2.hex b/test/util/data/txcreatemultisig2.hex
index 07835c54d3e1612b3bbe24eb8d4156902bd80046..f6669a0cdf7fcd9230b01cf622f9ff598187415c 100644
--- a/test/util/data/txcreatemultisig2.hex
+++ b/test/util/data/txcreatemultisig2.hex
@@ -1 +1 @@
-01000000000100e1f5050000000017a9141c6fbaf46d64221e80cbae182c33ddf81b9294ac8700000000
+01000000000100407a10f35a000017a9141c6fbaf46d64221e80cbae182c33ddf81b9294ac8700000000
diff --git a/test/util/data/txcreatemultisig2.json b/test/util/data/txcreatemultisig2.json
index 9b0cfdfd8cb7c2ca71f954e887108e21ddf18e84..ed2c5b68b0d70a9e3c4ab16da940b59e154c0a47 100644
--- a/test/util/data/txcreatemultisig2.json
+++ b/test/util/data/txcreatemultisig2.json
@@ -1,6 +1,6 @@
 {
-    "txid": "0d861f278a3b7bce7cb5a88d71e6e6a903336f95ad5a2c29b295b63835b6eee3",
-    "hash": "0d861f278a3b7bce7cb5a88d71e6e6a903336f95ad5a2c29b295b63835b6eee3",
+    "txid": "7bacb3311f0f9ec009c45f4037f3387aeeab33ec711fe96e37c15c509735d119",
+    "hash": "7bacb3311f0f9ec009c45f4037f3387aeeab33ec711fe96e37c15c509735d119",
     "version": 1,
     "size": 42,
     "locktime": 0,
@@ -8,7 +8,7 @@
     ],
     "vout": [
         {
-            "value": 1000000.00,
+            "value": 1000000.00000000,
             "n": 0,
             "scriptPubKey": {
                 "asm": "OP_HASH160 1c6fbaf46d64221e80cbae182c33ddf81b9294ac OP_EQUAL",
@@ -21,5 +21,5 @@
             }
         }
     ],
-    "hex": "01000000000100e1f5050000000017a9141c6fbaf46d64221e80cbae182c33ddf81b9294ac8700000000"
+    "hex": "01000000000100407a10f35a000017a9141c6fbaf46d64221e80cbae182c33ddf81b9294ac8700000000"
 }
diff --git a/test/util/data/txcreateoutpubkey1.json b/test/util/data/txcreateoutpubkey1.json
index 55e9de631a102303607412b2f00c481a2541f2f6..f70c6b8845fe76baf66ba4f4bff33049ff280928 100644
--- a/test/util/data/txcreateoutpubkey1.json
+++ b/test/util/data/txcreateoutpubkey1.json
@@ -8,7 +8,7 @@
     ],
     "vout": [
         {
-            "value": 0.00,
+            "value": 0.00000000,
             "n": 0,
             "scriptPubKey": {
                 "asm": "02a5613bd857b7048924264d1e70e08fb2a7e6527d32b7ab1bb993ac59964ff397 OP_CHECKSIG",
diff --git a/test/util/data/txcreatescript1.json b/test/util/data/txcreatescript1.json
index 857df3ec9eb85ccf81bfd7149823377067181f87..c8389e3ccc080a15b2fa768fb83de29efae2ba5a 100644
--- a/test/util/data/txcreatescript1.json
+++ b/test/util/data/txcreatescript1.json
@@ -8,7 +8,7 @@
     ],
     "vout": [
         {
-            "value": 0.00,
+            "value": 0.00000000,
             "n": 0,
             "scriptPubKey": {
                 "asm": "OP_DROP",
diff --git a/test/util/data/txcreatescript2.json b/test/util/data/txcreatescript2.json
index f8c295c52929af52905c80a11df97b70bf8357e5..4735631d54e7324b2f79793963df99bc3c2fbd8b 100644
--- a/test/util/data/txcreatescript2.json
+++ b/test/util/data/txcreatescript2.json
@@ -8,7 +8,7 @@
     ],
     "vout": [
         {
-            "value": 0.00,
+            "value": 0.00000000,
             "n": 0,
             "scriptPubKey": {
                 "asm": "OP_HASH160 71ed53322d470bb96657deb786b94f97dd46fb15 OP_EQUAL",
diff --git a/test/util/data/txcreatesignv1.hex b/test/util/data/txcreatesignv1.hex
index 4d4f96a26fbba6126f7909fbb65a8ba2f6fcff20..06efd897ed917588d4b54356be30fffd9f7b4ad1 100644
--- a/test/util/data/txcreatesignv1.hex
+++ b/test/util/data/txcreatesignv1.hex
@@ -1 +1 @@
-01000000018594c5bdcaec8f06b78b596f31cd292a294fd031e24eec716f43dac91ea7494d000000008a473044022042da3e29f29518ebcf1fa40a98a285e91e2dfcf472f5e3cde71875650009f2f1022011aab4c37d237c6e31d6c81ad099b259715ecd9ea83d0b302e6938386f35831101410479be667ef9dcbbac55a06295ce870b07029bfcdb2dce28d959f2815b16f81798483ada7726a3c4655da4fbfc0e1108a8fd17b448a68554199c47d08ffb10d4b8ffffffff01a0860100000000001976a9145834479edbbe0539b31ffd3a8f8ebadc2165ed0188ac00000000
+01000000018594c5bdcaec8f06b78b596f31cd292a294fd031e24eec716f43dac91ea7494d000000008a473044022042da3e29f29518ebcf1fa40a98a285e91e2dfcf472f5e3cde71875650009f2f1022011aab4c37d237c6e31d6c81ad099b259715ecd9ea83d0b302e6938386f35831101410479be667ef9dcbbac55a06295ce870b07029bfcdb2dce28d959f2815b16f81798483ada7726a3c4655da4fbfc0e1108a8fd17b448a68554199c47d08ffb10d4b8ffffffff0100e87648170000001976a9145834479edbbe0539b31ffd3a8f8ebadc2165ed0188ac00000000
diff --git a/test/util/data/txcreatesignv1.json b/test/util/data/txcreatesignv1.json
index 7fcc15c2ef3764d7f130da110c5650277737c0ce..278dd5287f52f87be4df6c67be75bb8269715edf 100644
--- a/test/util/data/txcreatesignv1.json
+++ b/test/util/data/txcreatesignv1.json
@@ -1,6 +1,6 @@
 {
-    "txid": "bea13b5f794dc351d1d890b525b0af423c49ab49c667c8fd95cbba2fb283cd5e",
-    "hash": "bea13b5f794dc351d1d890b525b0af423c49ab49c667c8fd95cbba2fb283cd5e",
+    "txid": "14ef38e9958fbfc28509ce54a2f301808064c68c07963c1f9e0b1838942e5a73",
+    "hash": "14ef38e9958fbfc28509ce54a2f301808064c68c07963c1f9e0b1838942e5a73",
     "version": 1,
     "size": 223,
     "locktime": 0,
@@ -17,7 +17,7 @@
     ],
     "vout": [
         {
-            "value": 1000.00,
+            "value": 1000.00000000,
             "n": 0,
             "scriptPubKey": {
                 "asm": "OP_DUP OP_HASH160 5834479edbbe0539b31ffd3a8f8ebadc2165ed01 OP_EQUALVERIFY OP_CHECKSIG",
@@ -30,5 +30,5 @@
             }
         }
     ],
-    "hex": "01000000018594c5bdcaec8f06b78b596f31cd292a294fd031e24eec716f43dac91ea7494d000000008a473044022042da3e29f29518ebcf1fa40a98a285e91e2dfcf472f5e3cde71875650009f2f1022011aab4c37d237c6e31d6c81ad099b259715ecd9ea83d0b302e6938386f35831101410479be667ef9dcbbac55a06295ce870b07029bfcdb2dce28d959f2815b16f81798483ada7726a3c4655da4fbfc0e1108a8fd17b448a68554199c47d08ffb10d4b8ffffffff01a0860100000000001976a9145834479edbbe0539b31ffd3a8f8ebadc2165ed0188ac00000000"
+    "hex": "01000000018594c5bdcaec8f06b78b596f31cd292a294fd031e24eec716f43dac91ea7494d000000008a473044022042da3e29f29518ebcf1fa40a98a285e91e2dfcf472f5e3cde71875650009f2f1022011aab4c37d237c6e31d6c81ad099b259715ecd9ea83d0b302e6938386f35831101410479be667ef9dcbbac55a06295ce870b07029bfcdb2dce28d959f2815b16f81798483ada7726a3c4655da4fbfc0e1108a8fd17b448a68554199c47d08ffb10d4b8ffffffff0100e87648170000001976a9145834479edbbe0539b31ffd3a8f8ebadc2165ed0188ac00000000"
 }
diff --git a/test/util/data/txcreatesignv2.hex b/test/util/data/txcreatesignv2.hex
index a102a6774272e8854c6e915dc634df3c9ecbb2ee..87a25b39f31f548a906a30698f6c475f24ac507b 100644
--- a/test/util/data/txcreatesignv2.hex
+++ b/test/util/data/txcreatesignv2.hex
@@ -1 +1 @@
-02000000018594c5bdcaec8f06b78b596f31cd292a294fd031e24eec716f43dac91ea7494d000000008a47304402205338be788a5c8ec16624f60c244927054e7b07bc8e540b14b1f4a16842ddefd102204e4db3e867caa416a6ec592addc30ba379c1e899913caf97214ed3e0aa3c866a01410479be667ef9dcbbac55a06295ce870b07029bfcdb2dce28d959f2815b16f81798483ada7726a3c4655da4fbfc0e1108a8fd17b448a68554199c47d08ffb10d4b8ffffffff01a0860100000000001976a9145834479edbbe0539b31ffd3a8f8ebadc2165ed0188ac00000000
+02000000018594c5bdcaec8f06b78b596f31cd292a294fd031e24eec716f43dac91ea7494d000000008a47304402205338be788a5c8ec16624f60c244927054e7b07bc8e540b14b1f4a16842ddefd102204e4db3e867caa416a6ec592addc30ba379c1e899913caf97214ed3e0aa3c866a01410479be667ef9dcbbac55a06295ce870b07029bfcdb2dce28d959f2815b16f81798483ada7726a3c4655da4fbfc0e1108a8fd17b448a68554199c47d08ffb10d4b8ffffffff0100e87648170000001976a9145834479edbbe0539b31ffd3a8f8ebadc2165ed0188ac00000000
